libs/langchain/langchain/chains/structured_output/base.py

```diff
@@ -0,0 +1,313 @@
+
+                llm = ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0)
+                system = '''You are a world class assistant for extracting information in structured JSON formats. \
+                
+                Extract a valid JSON blob from the user input that matches the following JSON Schema:
+                
+                {output_schema}'''
+                prompt = ChatPromptTemplate.from_messages(
+                    [
+                        ("system", system),
+                        ("human", "{input}"),
+                    ]
+                )
+                chain = create_structured_output_runnable(Dog, llm, prompt, mode="openai-json")
+                chain.invoke({"input": "Harry was a chubby brown beagle who loved chicken"})
+    """  # noqa: E501
+    if mode == "openai-functions":
+        return _create_openai_functions_structured_output_runnable(
+            output_schema,
+            llm,
+            prompt,
+            output_parser=output_parser,
+            enforce_single_function_usage=enforce_single_function_usage,
+            **kwargs,
+        )
+    elif mode == "openai-json":
+        return _create_openai_json_runnable(
+            output_schema, llm, prompt, output_parser=output_parser, **kwargs
+        )
+    else:
+        raise ValueError(
+            f"Invalid mode {mode}. Expected one of 'openai-functions', "
+            f"'openai-json'."
+        )
+
+
+def get_openai_output_parser(
+    functions: Sequence[Union[Dict[str, Any], Type[BaseModel], Callable]],
+) -> Union[BaseOutputParser, BaseGenerationOutputParser]:
+    """Get the appropriate function output parser given the user functions.
+
+    Args:
+        functions: Sequence where element is a dictionary, a pydantic.BaseModel class,
+            or a Python function. If a dictionary is passed in, it is assumed to
+            already be a valid OpenAI function.
+
+    Returns:
+        A PydanticOutputFunctionsParser if functions are Pydantic classes, otherwise
+            a JsonOutputFunctionsParser. If there's only one function and it is
+            not a Pydantic class, then the output parser will automatically extract
+            only the function arguments and not the function name.
+    """
+    function_names = [convert_to_openai_function(f)["name"] for f in functions]
+    if isinstance(functions[0], type) and issubclass(functions[0], BaseModel):
+        if len(functions) > 1:
+            pydantic_schema: Union[Dict, Type[BaseModel]] = {
+                name: fn for name, fn in zip(function_names, functions)
+            }
+        else:
+            pydantic_schema = functions[0]
+        output_parser: Union[
+            BaseOutputParser, BaseGenerationOutputParser
+        ] = PydanticOutputFunctionsParser(pydantic_schema=pydantic_schema)
+    else:
+        output_parser = JsonOutputFunctionsParser(args_only=len(functions) <= 1)
+    return output_parser
+
+
+def _create_openai_json_runnable(
```

Comment:
main addition

Reply:
Functions API is deprecated, we should be using tools for new stuff

Reply:
this isnt using functions